community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/ParallelSort.java

```diff
@@ -215,23 +216,24 @@ else if ( Utils.unsignedCompare( right, pivot, CompareType.GE ) )
         return partingIndex;
     }
 
-    private void recursiveQsort( int start, int end )
+    private void recursiveQsort( int start, int end, Random random )
     {
         if ( end - start < 2 )
         {
             return;
         }
-        //choose the middle value
-        int pivot = start + ((end - start) / 2);
+        // choose a random pivot between start and end
+        int pivot = start + random.nextInt( end - start );
 
         pivot = partition( start, end, pivot );
 
-        recursiveQsort( start, pivot );
-        recursiveQsort( pivot + 1, end );
+        recursiveQsort( start, pivot, random );
+        recursiveQsort( pivot + 1, end, random );
     }
 
     private class SortWorker extends Thread
     {
+        private final Random random = new Random();
```

Comment:
You can alternatively grab a `ThreadLocalRandom` in the `run()` method – it uses a faster algorithm because it knows it doesn't have to worry about contention.